### PR TITLE
OpenStack: allow cluster-api-provider-openstack to read configmaps 

### PR DESCRIPTION
**Pull request by iamemilio**
CAPO needs to read the ca-cert from the openshift-config:cloud-provider-config config map, and needs RBAC access : https://github.com/openshift/cluster-api-provider-openstack/pull/79/files

This change is part of self signed certs effort outlined here: https://docs.google.com/document/d/1bKXArO_dtIOQMVqV35GTmUdzoRQAgUqIVMbO9kpfknI/edit?folder=0AFMASrm3oxfNUk9PVA#heading=h.8jarwghidch7

Other relevant ssc changes:
Ensure kubelet doesnt fail on startup: https://github.com/openshift/machine-config-operator/pull/1392/files
Generate key in configmap with ca-cert: https://github.com/openshift/installer/pull/2932/files

### CODE REVIEW

**Comment by iamemilio** [comment]
@sdodson @enxebre Is this going to fly?

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 1]
```diff
@@ -116,6 +116,13 @@ rules:
       - list
       - watch
 
+  - apiGroups:
+      - ""
+    resoureces:
```
Comment (mandre) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: s/resoureces/resources/

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 2]
```diff
@@ -116,6 +116,13 @@ rules:
...
+      - configmap
+    verbs:
+      - get
```
Comment (mandre) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: Perhaps better to also grant list and watch permissions.

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 3]
```diff
@@ -332,6 +332,21 @@ subjects:
     name: prometheus-k8s
     namespace: openshift-monitoring
 
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: cloud-provider-configmap-reader
+  namespace: openshift-machine-api
```
Comment (JoelSpeed) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: The rolebinding needs to be in the namespace that the role is in, so this needs to be `openshift-config`
```suggestion
  namespace: openshift-config
```

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 4]
```diff
@@ -356,3 +371,18 @@ rules:
...
+metadata:
+  name: cloud-provider-configmap-reader
+  namespace: openshift-config
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - configmap
+    verbs:
+      - get
```
Comment (JoelSpeed) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: Nit: missing new line at the end of the file

Should probably add `list` and `watch` to this too

Is the ConfigMap always named the same, or can it vary depending on cluster names etc? You _could_ be extra strict and restrict it to the particular ConfigMap name if that is statis

Comment (iamemilio) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: Will do, and I it should always have the same name

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 5]
```diff
@@ -332,6 +332,21 @@ subjects:
...
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: cloud-provider-configmap-reader
+  namespace: openshift-config
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: cloud-provider-configmap-reader
```
Comment (enxebre) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: In terms of naming `cloud-provider` is circunstancial here, this role gives you ability to read all configmaps in the namespace regardless a consumer uses it for reading the cloud-provider or something else.
Also I wonder if this should reflect ownership and be named after `machine-api-controllers-*`? cc @JoelSpeed @bison wdyt?

Also can you please add this to the related resources https://github.com/openshift/machine-api-operator/blob/1c9807c657875c76f2b0e7f93d2075cee5e6af10/pkg/operator/status.go#L182-L214?

Comment (iamemilio) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: Is there any way to narrow things down to just the singular configmap?

**Comment by JoelSpeed** [comment]
Re the naming and possibly restricting, if it were restricted to the single resource, I'd suggest `machine-api-controllers-cloud-provider-configmap-reader` as a verbose but accurate name, I can definitely see the value in naming it after `machine-api-controllers` otherwise is does look like a bit of a random role if you happen to be looking at the roles in the `openshift-config` namespace

Sticking with all configmaps maybe `machine-api-controllers-configmap-reader`?

I had a look at the openstack code and so if we want to restrict, should be fairly easy, just replace the rules with:
```
- apiGroups:
  - ""
  resources:
  - configmap
  resourceNames:
  - cloud-provider-config
  verbs:
  - get
```

**Comment by enxebre** [comment]
Commit message is a bit confusing to me `allow machine-api-service account to get cloud-config`, the pr actually `allow machine-api-controllers serviceAccount to fetch configmap in the openshift-config namespace`. Also can you include context on why and references to counterpart PRs in the commit description i.e `git commit -m"what it does" -m"why, e.g to read the cloud provider config for reasons..."`
Is this based on an https://github.com/openshift/enhancements/tree/master/enhancements proposal?

File: install/0000_30_machine-api-operator_09_rbac.yaml (hunk 5 above)
Comment (iamemilio) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: @enxebre What did you want me to add in there? 
```json
...
               {
			Group:    "rbac.authorization.k8s.io",
			Resource: "roles",
			Name:     "machine-api-controllers",
		}
}
```

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 6]
```diff
@@ -356,3 +371,20 @@ rules:
       - get
       - list
       - watch
+
+---
+# Roles needed by cluster-api in order to read openshift-config:cloud-provider-config configmap
```
Comment (enxebre) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: nit: replace cluster-api with machine-api

File: pkg/operator/status.go [hunk 7]
```diff
@@ -213,6 +213,11 @@ func (optr *Operator) getOrCreateClusterOperator() (*osconfigv1.ClusterOperator,
 			Resource: "clusterroles",
 			Name:     "machine-api-controllers",
 		},
+		{
+			Group:    "rbac.authorization.k8s.io",
+			Resource: "roles",
+			Name:     "machine-api-controllers",
```
Comment (enxebre) [inline pkg/operator/status.go]: there's no role with this name in current PR

File: install/0000_30_machine-api-operator_09_rbac.yaml (hunk 6 above)
Comment (enxebre) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: can you add a comment referencing where this is coming from i.e https://github.com/openshift/installer/blob/a76a6b34092d99835d0e947ae9343d231f38f7e1/pkg/asset/manifests/cloudproviderconfig.go#L80

I'm not totally sure it's fair to assume this name or should be inferred from https://github.com/openshift/installer/blob/master/pkg/asset/manifests/infrastructure.go#L167

File: pkg/operator/status.go (hunk 7 above)
Comment (iamemilio) [reply pkg/operator/status.go]: whoops :)

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 8]
```diff
@@ -332,6 +332,21 @@ subjects:
     name: prometheus-k8s
     namespace: openshift-monitoring
 
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: machine-api-controllers
```
Comment (michaelgugino) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: There is already a RoleBinding with this name.  The name here should be cloud-provider-config-reader I think, or possibly something like 'machine-api-cloud-provider-config-reader' so there aren't collisions with other components.

Comment (iamemilio) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: There was a prior suggestion to have "machine-api" in the name to show ownership, so I will go with "machine-api-cloud-provider-config-reader". Thanks!

**Comment by eparis** [comment]
@michaelgugino did you want to /approve as well   :)

File: install/0000_30_machine-api-operator_09_rbac.yaml (hunk 8 above)
Comment (enxebre) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: I'd rather keep the name as `machine-api-controllers` so we are consistent across  with the existing ones and we can clearly identify owned by this component in each namespace, not a blocker...

**Pull request merged**